R/run-nmtran.R

```diff
@@ -0,0 +1,330 @@
+
+  # Copy dataset & overwrite $DATA record of new model
+  # NM-TRAN will error if data cannot be found
+  if(fs::file_exists(data_path)){
+    file.copy(data_path, temp_folder)
+    # overwrite $DATA record of new model
+    modify_data_path_ctl(nmtran_mod, data_path = basename(data_path))
+  }
+
+  # Run NM-TRAN
+  nmtran_results <- c(
+    list(
+      nmtran_exe = nmtran_specs$nmtran_exe,
+      nmtran_presort_exe = nmtran_specs$nmtran_presort_exe,
+      nonmem_version = nmtran_specs$nonmem_version,
+      absolute_model_path = mod_path
+    ),
+    execute_nmtran(
+      nmtran_specs$nmtran_exe,
+      mod_path = basename(mod_path),
+      cmd_args = nmtran_specs$cmd_args,
+      nmtran_presort_exe = nmtran_specs$nmtran_presort_exe,
+      dir = temp_folder
+    )
+  )
+
+  # assign class and return
+  class(nmtran_results) <- c(NMTRAN_PROCESS_CLASS, class(nmtran_results))
+  return(nmtran_results)
+}
+
+
+#' Sets up `NM-TRAN` to run by identifying an `NM-TRAN` executable for a
+#'  provided `NONMEM` version and formatting arguments based on `.bbi_args`,
+#'  the corresponding `.mod` yaml file, and any `nmfe_options` defined in a
+#'  `bbi.yaml` file
+#'
+#' @inheritParams run_nmtran
+#' @keywords internal
+nmtran_setup <- function(
+    .mod = NULL,
+    .nonmem_version = NULL,
+    .bbi_args = NULL,
+    .config_path = NULL
+){
+  bbi_yaml_path <- get_bbi_yaml_path(.mod, .config_path = .config_path)
+  bbi_yaml <- yaml::read_yaml(bbi_yaml_path)
+
+  # Combine NONMEM submission args
+  #  - The main ones of interest are prdefault, tprdefault, and maxlim, which
+  #    impact the evaluation of `NM-TRAN`
+  #  - Priority: .bbi_args > model yaml > bbi.yaml
+  cmd_args <- parse_nmtran_args(
+    .mod, .bbi_args = .bbi_args, nmfe_options = bbi_yaml$nmfe_options
+  )
+
+  # Check nonmem version
+  nm_config <- bbi_yaml$nonmem
+  if (!is.null(.nonmem_version)) {
+    # check for valid version
+    if (!(.nonmem_version %in% names(nm_config))) {
+      rlang::abort(
+        c(
+          "Must specify a valid `.nonmem_version` for run_nmtran.",
+          "i" = glue("{bbi_yaml_path} contains the following options:"),
+          glue("`{paste(names(nm_config), collapse='`, `')}`")
+        )
+      )
+    }
+    default_nm <- nm_config[[.nonmem_version]]
+  }else{
+    # Look for default nonmem installation
+    default_nm <- purrr::keep(nm_config, function(nm_ver){
+      !is.null(nm_ver$default) & isTRUE(nm_ver$default)
+    })
+
+    if(length(default_nm) > 1){
+      nm_vers <- paste(names(default_nm), collapse = ", ")
+      rlang::abort(
+        c(
+          glue("Found multiple default NONMEM versions ({nm_vers}) at `{config_path}`"),
+          "i" = "Please ensure only one version is set to the default"
+        )
+      )
+    }else if(length(default_nm) == 1){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
```

Comment:
A `NULL` value for `nm_config` can reach here. (Yes, it'd be unlikely, either someone editing `bbi.yaml` directly or accidentally passing the path to another yaml.) In that case, `default_nm` will be `NULL`, the downstream handling will be incorrect, and you'll get a "character vector argument expected" error at the `basename(default_nm$home)` call on line 175.